src/util.jl

```diff
@@ -162,49 +162,101 @@ function solve_triangular_system(d::OrderedDict)
     OrderedDict{Symbol,Real}(zip(nms, data))
 end
 
+type TriangularSystemException <: Exception
+    missing
+end
 
 """
 Solves triangular system specified by incidence dictionary.
 
-Returns a Vector of keys specifying the order in which `dd` should be evaluated
-
 ```
-system = Dict(0=>[1,2], 1=>Int[], 2=>[1] )
+system = Dict(0=>[1,2], 1=>[], 2=>[1] )
 solve_dependency(system)
 ```
 
 or
 
 ```
-system = Dict(:x=>[:y,:z], :y=>Symbol[], :z=>[:y])
+system = Dict(:x=>[:y,:z], :y=>[], :z=>[:y] )
 solve_dependency(system)
 ```
+
+Optionally, one can add specify which subset of variables to solve for.
+Unrequired variables will be ommited in the solution.
+
 """
-function solve_dependency(dd::Dict{T,Set{T}}) where T # is hashable
-    solved = T[]
-    deps = deepcopy(dd)
-    p = length(deps)
-    it = 0
-    while length(deps) > 0 && it <= p
-        it += 1
-        for (k, dep) in deps
-            if length(dep) == 0
-                push!(solved, k)
-                pop!(deps, k)
-                for (l, ldeps) in deps
-                    if k in ldeps
-                        pop!(ldeps, k)
-                    end
+function solve_dependencies(deps::Associative{T,Set{T}}, needed=nothing) where T
+    solution = []
+    if needed == nothing
+        needed = Set(keys(deps))
+    else
+        needed = Set(needed)
+    end
+    while length(needed)>0
+        tt0 = (length(needed), length(solution))
+        if length(needed)==0
+            return solution
+        else
+            for k in needed
+                # check whether k is solved
+                if k in solution
+                    pop!(needed, k)
+                elseif issubset(deps[k], solution)
+                    push!(solution, k)
+                else
+                    needed = union(deps[k], needed)
                 end
             end
+            tt = (length(needed), length(solution))
+            if tt == tt0
+                mis = join([string(e) for e in needed], ", ")
+                exc = TriangularSystemException(mis)
+                throw(exc)
+            end
         end
     end
-    if it >= p+1
-        error("Non triangular system")
+    return solution
+end
+
+list_symbols(expr::Symbol) = list_symbols(:($expr*2))
```

Comment:
haha, is the `*2` just to convert this symbol into an expression?

Reply:
OK, I confess I was too lazy to look up the type of output list_symbols was returning... Will fix

Reply:
Now `list_symbols(expr::Symbol, other...) =  Dict{Symbol,Any}(:parameters=>[expr])`